File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/8815fe9150c4f82f49c27cd0403c8142.go
package main

import (
	"fmt"
	"math"
)

type Mint struct {
	v uint
	mod uint
}

func NewMint(val uint, mod uint) Mint {
	return Mint{v: val % mod, mod: mod}
}

func (m Mint) Add(other Mint) Mint {
	return NewMint(m.v + other.v, m.mod)
}

func (m Mint) Sub(other Mint) Mint {
	return NewMint(m.v + m.mod - other.v, m.mod)
}

func (m Mint) Mul(other Mint) Mint {
	return NewMint(m.v * other.v % m.mod, m.mod)
}

func (m Mint) Inv() Mint {
	return NewMint(Inverse(m.v, m.mod), m.mod)
}

func Inverse(a, mod uint) uint {
	m0, y, x := mod, uint(0), uint(1)
	if mod == 1 {
		return 0
	}
	for a > 1 {
		q := a / mod
		m0, a, mod = a, mod, a%mod
		y, x = x-y*q, m0-y*q
	}
	if x < 0 {
		x += m0
	}
	return x
}

func Popcount(u uint) uint {
	return uint(math.PopCount(uint64(u)))
}

func main() {
	const mod = 998244353
	var n, m int
	fmt.Scan(&n, &m)

	a := make([]bool, n)
	for i := 0; i < n; i++ {
		var x int
		fmt.Scan(&x)
		a[i] = x != 0
	}

	w := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&w[i])
	}

	zw, ow := 0, 0
	for i := 0; i < n; i++ {
		if a[i] {
			ow += w[i]
		} else {
			zw += w[i]
		}
	}
	s := zw + ow

	inv := make([]Mint, 2*m+1)
	for d := -m; d <= m; d++ {
		if s+d > 0 {
			inv[d+m] = NewMint(1, s+d)
		}
	}

	dp := make([][]Mint, m+1)
	for i := range dp {
		dp[i] = make([]Mint, 2)
	}
	dp[0][0] = NewMint(1, mod)

	for i := 0; i < m; i++ {
		ndp := make([][]Mint, m+1)
		for j := range ndp {
			ndp[j] = make([]Mint, 2)
		}
		for z := 0; z <= i; z++ {
			o := i - z
			wz := zw - z
			wo := ow + o
			if wz < 0 {
				continue
			}
			for dz := -1; dz <= 0; dz++ {
				nwz := wz + dz
				nz := z - dz
				if nwz < 0 {
					continue
				}
				eps := inv[wz+wo-s+m]
				p := Mint{v: 0, mod: mod}
				if dz == -1 {
					p = Mint{v: wz, mod: mod}.Mul(eps)
				} else {
					p = Mint{v: wo, mod: mod}.Mul(eps)
				}
				ndp[nz][-dz] = ndp[nz][-dz].Add(dp[z][-dz].Mul(p))
				ndp[nz][1+dz] = ndp[nz][1+dz].Add(p.Add(Mint{v: uint(2*dz + 1), mod: mod}.Mul(eps)).Mul(dp[z][1+dz]))
			}
		}
		dp = ndp
	}

	cz, co := Mint{v: 0, mod: mod}, Mint{v: 0, mod: mod}
	for i := 0; i <= m; i++ {
		cz = cz.Add(dp[i][0])
		co = co.Add(dp[i][1])
	}

	for i := 0; i < n; i++ {
		if a[i] {
			fmt.Println(w[i] * co.v)
		} else {
			fmt.Println(w[i] * cz.v)
		}
	}
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/C/8815fe9150c4f82f49c27cd0403c8142.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <limits.h>
#include <math.h>

typedef int32_t i32;
typedef int64_t i64;
typedef uint32_t u32;
typedef uint64_t u64;
typedef unsigned int uint;
typedef size_t usize;
typedef long long ll;
typedef unsigned long long ull;
typedef long double ld;

#define MOD 998244353

typedef struct {
    uint v;
} modint;

modint make(uint x) {
    modint m;
    m.v = x % MOD;
    return m;
}

modint add(modint m1, modint m2) {
    return make(m1.v + m2.v);
}

modint sub(modint m1, modint m2) {
    return make(m1.v + MOD - m2.v);
}

modint mul(modint m1, modint m2) {
    return make((ll)m1.v * m2.v % MOD);
}

modint inv(ll v) {
    if (v < 1000000) {
        // Small inverse logic can be implemented here if needed
        return make(1); // Placeholder
    }
    // Extended Euclidean algorithm for modular inverse
    ll a = v, m = MOD, m0 = m, y = 0, x = 1;
    if (m == 1) return make(0);
    while (a > 1) {
        ll q = a / m;
        ll t = m;
        m = a % m, a = t;
        t = y;
        y = x - q * y;
        x = t;
    }
    if (x < 0) x += m0;
    return make(x);
}

modint power(modint x, ull n) {
    modint ans = make(1);
    while (n) {
        if (n & 1) ans = mul(ans, x);
        x = mul(x, x);
        n >>= 1;
    }
    return ans;
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);
    
    bool *a = malloc(n * sizeof(bool));
    int *w = malloc(n * sizeof(int));
    
    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }
    
    for (int i = 0; i < n; i++) {
        scanf("%d", &w[i]);
    }
    
    int zw = 0, ow = 0;
    for (int i = 0; i < n; i++) {
        if (a[i]) ow += w[i];
        else zw += w[i];
    }
    
    int s = zw + ow;
    modint *inv = malloc((2 * m + 1) * sizeof(modint));
    
    for (int d = -m; d <= m; d++) {
        if (s + d > 0) inv[d + m] = inv(s + d);
    }
    
    modint **dp = malloc((m + 1) * sizeof(modint *));
    for (int i = 0; i <= m; i++) {
        dp[i] = malloc(2 * sizeof(modint));
        dp[i][0] = make(0);
        dp[i][1] = make(0);
    }
    dp[0][0] = make(1);
    
    for (int i = 0; i < m; i++) {
        modint **ndp = malloc((m + 1) * sizeof(modint *));
        for (int j = 0; j <= m; j++) {
            ndp[j] = malloc(2 * sizeof(modint));
            ndp[j][0] = make(0);
            ndp[j][1] = make(0);
        }
        
        for (int z = 0; z <= i; z++) {
            int o = i - z;
            int wz = zw - z;
            int wo = ow + o;
            if (wz < 0) continue;
            for (int dz = -1; dz <= 0; dz++) {
                int nwz = wz + dz;
                int nz = z + (-dz);
                if (nwz < 0) continue;
                modint eps = inv[wz + wo - s + m];
                modint p = mul(dz == -1 ? make(wz) : make(wo), eps);
                ndp[nz][-dz] = add(ndp[nz][-dz], mul(dp[z][-dz], p));
                ndp[nz][1 + dz] = add(ndp[nz][1 + dz], add(p, mul(make(2 * dz + 1), eps)));
            }
        }
        for (int j = 0; j <= m; j++) {
            free(dp[j]);
        }
        free(dp);
        dp = ndp;
    }
    
    modint cz = make(0), co = make(0);
    for (int i = 0; i <= m; i++) {
        cz = add(cz, dp[i][0]);
        co = add(co, dp[i][1]);
    }
    
    for (int i = 0; i < n; i++) {
        printf("%lld\n", (a[i] == 0 ? cz.v : co.v) * w[i]);
    }
    
    for (int i = 0; i <= m; i++) {
        free(dp[i]);
    }
    free(dp);
    free(inv);
    free(a);
    free(w);
    
    return 0;
}
